### PR TITLE
Reorganize gc.c to prepare for thread-safe variant, with thread context hack.

### PR DESCRIPTION
**Pull request by ArchRobison**
This is a continuation of #10724.  It's a new PR because Github does [not allow](https://github.com/isaacs/github/issues/361)  reopening a pull request after a forced push, which I learned the hard way :frowning: 

The goal of this PR is still the same: modify `gc.c` in ways to prepare for making it thread-safe later.   All changes in this PR are intended to be cosmetic and not impact functionality or performance.

The improvement over #10724 is the use of macros `FOR_CURRENT_HEAP` and `FOR_EACH_HEAP` for establishing thread context.  In the thread-safe version, these will look something like:

```
#define FOR_CURRENT_HEAP {jl_thread_heap_t* current_heap = jl_thread_heap;
#define END }
#define FOR_EACH_HEAP for( int t=jl_n_threads; --t>=0; )  { jl_thread_heap_t* current_heap = jl_all_heaps[t];
/*}*/
#define HEAP(x) (current_heap->x)
```

In the current PR, these macros devolve in a way that retains the current logic, albeit with a dummy variable `current_heap` that is used to check that the macros are being used where needed (albeit no checking for _correct_ use.)  Any modern compiler will optimize away the dummy variable.

The other change over #10724 is that I put the finalizer stuff back in its original position.  Because one thread can set up a finalizer, but another thread can ask for it to be executed, correct implementation is going to require more than just per-thread lists, so that's a battle for another day.

### CODE REVIEW

File: src/gc.c [hunk 1]
```diff
@@ -665,59 +721,65 @@ static inline int maybe_collect(void)
...
+            if (gc_marked(jl_astaggedvalue(wr))) {
+                // weakref itself is alive
+                if (!gc_marked(jl_astaggedvalue(wr->value)))
+                    wr->value = (jl_value_t*)jl_nothing;
+                n++;
+            }
+            else {
+                ndel++;
+            }
+        } while ((n < l-ndel) && SWAP_wr(lst[n],lst[n+ndel]));
```
Comment (StefanKarpinski) [inline src/gc.c]: The tabs here may make the build fail.

Comment (ArchRobison) [inline src/gc.c]: I was wondering what the extra green meant.  I guessed tabs, but I couldn't  find them in my working copy.  I'll poke around to figure out how to eradicate them.

Comment (StefanKarpinski) [inline src/gc.c]: This out to work:

```sh
perl -i -ple 's/\t/    /g'
```

Comment (ArchRobison) [inline src/gc.c]: The problem is convincing Github that the tabs are not there.   When I check out my branch, `src/gc.c` has no tabs.

Comment (Keno) [inline src/gc.c]: Are those actually hard tabs? I believe github also displays it like this for soft tabs.

Comment (ArchRobison) [inline src/gc.c]: What is a soft tab?

Comment (Keno) [inline src/gc.c]: Spaces.

Comment (ArchRobison) [inline src/gc.c]: What's puzzling me is why it's displaying those lines differently than many others that I also indented by 4 spaces.

Comment (Keno) [inline src/gc.c]: GitHub can't always detect changes within a line. E.g. below where you added a static declaration, the same green box exists.

**Comment by ArchRobison** [comment]
Patch is 100% tab-free ingredients.  @carnaval - okay to commit?

**Comment by carnaval** [comment]
LGTM. The macros don't even look that hacky to me, it's pretty much all we need inside gc as far as I can tell. This + a real implementation of the thread contexts + a barrier around the collection and we should have something somewhat working.

**Pull request merged**